onboarding/mlsys/2-autodiff.md

```diff
@@ -0,0 +1,399 @@
+# Part 2: Autodifferentiation and Optimizers
+
+In "Part 1: Calculating Gradients and Averaging Them Together", we learned how
+gradient descent works mathematically.
+
+In this exercise, you'll learn how an actual deep learning library implements
+gradient descent.  You'll also learn about what optimizers are in a deep
+learning framework, and you'll implement one yourself, in both PyTorch and
+Tensorflow.  And lastly, you'll
+
+## Autodifferentiation
+
+Autodifferentiation is a mechanism by which gradients may be calculated
+symbolically for a model of arbitrary size and complexity by repeated
+applications of the product rule.
+
+You may remember from calculus class that the chain rule says that for
+function $h$ composed of functions $f$ and $g$:
+
+```math
+h(x) = f(g(x))
+```
+
+then the derivative of $h$ is:
+
+```math
+h'(x) = f'(g(x)) \cdot g'(x)
+```
+
+The chain rule applies to derivatives of nested functions.  If we think of
+successive layers of a model as nested functions, then we can imagine the
+following:
+
+- $x$ is the input to the model
+- $g$ is the composition of all previous layers of the model
+- $g(x)$ is the output of the previous layer
+- $f$ is the current layer of the model
+- $h$ is the composition of all previous layers and also the current one
+
+Now let's replace the term $f'(g(x))$ with an arbitrary variable name $Z$:
+
+```math
+h'(x) = Z \cdot g'(x)
+```
+
+We can say that $Z$ must be a value that, multiplied by the gradient of previous
+layers ($g'(x)$), results in the gradient including the current layer ($h'(x)$).
+
+If $Z$ were particularly easy to calculate, it would be a way to calculate the
+total gradient incrementally, by calculating the equivalent $Z$ term for each
+layer and multiplying them together.  But is $Z$ easy to calculate for a layer?
+
+The answer turns out to be "yes".  $Z$ is the result of feeding the output of
+the previous layer $g(x)$, which is easy to compute, into $f'$, the derivative
+of the operator $f$, which is often trivial to obtain (for example, if $f(n) =
+n^2$ then $f'(n) = 2n$).
```

Comment:
Personal preference, but I found this section a little confusing. Seems like it's asking the reader to like pack `f'(g)` into whatever `Z` is, forget about `Z`s origin, and then rediscover that it is just `f'(g)` again.

I think you may be going for like a recursive picture here, which is the more standard viewpoint.  There is some "upstream gradient" which has been computed already and passed to us, which here is like the derivative of the loss with respect to `f`: `dL/df`.  If we are interested in the derivative of the loss with respect to `g`, then you just need to know `f'(g)`, from which you get `dL/dg = dL/df * f'(g)`.  And then you probably need derivatives of tensors from lower layers still, for which you can now start with the upstream gradient `dL/dg` and so on and so forth.